### PR TITLE
Add UppercaseStringProcessor

### PR DESCRIPTION
**Pull request by dapowers87**
The first round of string processing to be integrated to the project, the UppercaseStringProcessor.

### Description
Includes the UppercaseStringProcessor, its base classes, and config classes, and conversion logic.
 
### Issues Resolved
Partially #697 
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/WithKeysProcessor.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    public WithKeysProcessor(final PluginMetrics pluginMetrics, final WithKeysProcessorConfig config) {
+        super(pluginMetrics);
+        this.keys = config.getWithKeys();
+    }
+
+    @Override
+    protected void performStringAction(Event recordEvent) {
+        for(String key : keys) {
+            final Object value = recordEvent.get(key, Object.class);
+            if(value instanceof String) {
```
Comment (sbayer55) [inline data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/WithKeysProcessor.java:25]: It looks like your method checks for a value of string to skip a value that is an unexpected type. If the key is present with a null value it looks like an unhandled runtime exception will still be thrown by `JacksonEvent`. Not sure if this is intended.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/MutateMapper.java [hunk 2]
```diff
@@ -54,11 +55,13 @@ public RenameCopyConfig(String fromKey, String toKey) {
...
                 });
             } else if(Objects.equals(name, "remove_field")) {
-                deletes.add((ArrayList<String>)attr.getAttributeValue().getValue());
+                deletes.addAll((ArrayList<String>)attr.getAttributeValue().getValue());
             } else if(Objects.equals(name, "copy")) {
                 ((Map<String, String>)attr.getAttributeValue().getValue()).entrySet().forEach(entry -> {
                     copies.add(new RenameCopyConfig(entry.getKey(), entry.getValue()));
                 });
+            } else if(Objects.equals(name, "uppercase")) {
+                uppercases.addAll((ArrayList<String>)attr.getAttributeValue().getValue());
```
Comment (dinujoh) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/MutateMapper.java:64]: This class is doing multiple things and not following single responsibility principle. Consider creating interface and implementation for each string processor in separate class to make this class more maintainable.

You can have use DI to inject Map of StringProcessorType -> Implementation. This class can lookup the Map for StringProcessor type and call the interface method. This will keep the code more logical and maintainable.

**Review by dinujoh — CHANGES_REQUESTED** [review]

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/MutateMapper.java (hunk 2 above)
Comment (dapowers87) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/MutateMapper.java:64]: Agreed. I mentioned changing this class to make it more like how you are saying during standup, but I want to get a sample processor and the base structure out for @sshivanii to build off of first.

File: data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/UppercaseStringProcessor.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import com.amazon.dataprepper.model.annotations.DataPrepperPlugin;
+import com.amazon.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import com.amazon.dataprepper.model.event.Event;
+import com.amazon.dataprepper.model.processor.Processor;
+import java.util.Locale;
+
+@DataPrepperPlugin(name = "uppercase_string", pluginType = Processor.class, pluginConfigurationType = UppercaseStringProcessorConfig.class)
+public class UppercaseStringProcessor extends WithKeysProcessor {
+    @DataPrepperPluginConstructor
+    public UppercaseStringProcessor(final PluginMetrics pluginMetrics, final UppercaseStringProcessorConfig config) {
```
[the remark this answers is not in the document]
Comment (dapowers87) [reply data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/UppercaseStringProcessor.java:18]: It can, I felt that creating an `UppercaseStringProcessorConfig`, albeit blank, was a good route to lay the groundwork for any enhancements down the road.

File: data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/WithKeysProcessor.java (hunk 1 above)
Comment (sbayer55) [reply data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/WithKeysProcessor.java:25]: After checking this scenario it looks like this in handled by jackson

File: data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/UppercaseStringProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbayer55) [reply data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/UppercaseStringProcessor.java:18]: Its probably best to remove the unused class for now. We can always add it back as needed.

File: data-prepper-plugins/mutate-string-processors/src/test/java/com/amazon/dataprepper/plugins/processor/mutatestring/UppercaseStringProcessorTests.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class UppercaseStringProcessorTests {
```
Comment (sbayer55) [inline data-prepper-plugins/mutate-string-processors/src/test/java/com/amazon/dataprepper/plugins/processor/mutatestring/UppercaseStringProcessorTests.java:30]: Thanks for including a comprehensive set of test cases.

**Review by sbayer55 — CHANGES_REQUESTED** [review]
Thanks for putting this together. After a quick updated based on the comments it's good to go.

**Review by sbayer55 — DISMISSED** [review]

File: data-prepper-plugins/mutate-string-processors/src/test/java/com/amazon/dataprepper/plugins/processor/mutatestring/UppercaseStringProcessorTests.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+        assertThat(editedRecords.get(0).getData().containsKey("message"), is(true));
+        assertThat(editedRecords.get(0).getData().get("message", Object.class), equalTo(null));
+    }
+
+    @Test
+    public void testValueIsObjectUppercaseStringProcessor() {
+        final UppercaseStringProcessor processor = createObjectUnderTest();
+        final TestObject testObject = new TestObject();
+        testObject.a = "MSG";
```
Comment (dinujoh) [inline data-prepper-plugins/mutate-string-processors/src/test/java/com/amazon/dataprepper/plugins/processor/mutatestring/UppercaseStringProcessorTests.java:106]: nit: consider making this lowercase since this class is testing for Uppercase strings even though this is not related.

**Review by sbayer55 — APPROVED** [review]
Consider adding a follow up PR adding javadocs to the Processor classes.

File: data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/WithKeysProcessorConfig.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    private List<String> withKeys;
+
+    @Override
+    public List<String> getIterativeConfig() {
+        return withKeys;
+    }
+
+    public List<String> getWithKeys() {
+        return withKeys;
+    }
```
Comment (asifsmohammed) [inline data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/WithKeysProcessorConfig.java:27]: It looks like these two getters are redundant.

Comment (dapowers87) [reply data-prepper-plugins/mutate-string-processors/src/main/java/com/amazon/dataprepper/plugins/processor/mutatestring/WithKeysProcessorConfig.java:27]: They are but I want users to have access to the keys naturally without having to dig in and figure out that `getIterativeConfig` is the keeper of the keys.

**Review by asifsmohammed — APPROVED** [review]
